### PR TITLE
Change log formatter to logfmt

### PR DESCRIPTION
**Pull request by kzemek**
This PR addresses #1248

It changes lager formatter to logfmt-compatible one. The formatter works in conjunction with lager's metadata, so logged tags can be added per-process as in this pull request's ejabberd_c2s.erl . We still require discipline to write logs in `task=adding_things msg="what's going on"` format, though (discuss in #1248 ?).

### CODE REVIEW

**Comment by kzemek** [comment]
Example log message: `app=mongooseim severity=debug node=mongooseim@localhost pid=<0.1105.0> line=60 function=update module=shaper application=ejabberd Tokens: 209 (+90,-194), delay: 0 ms`

**Comment by michalwski** [comment]
The use of lager meta data is very cool, having f.e user's JID for every log printed from c2s will much helpful. I home some questions comments to this formatter:

1. There is no date and time when the log message was added. I found this very important especially when doing post-mortem analyzation.
2. Proposed changes are only about console backend. That's fine for debugging, in production logs are rather written to the ejabberd.log file or even syslog backend and stored in a service for logs like `Greylog` or `Splunk`. I'd be good to have the same log format in all backends (default backends being console and file).
3. I'm personally in favour of using `level=info` instead of `severity=level` and heaving this part of log message at the beginning (just after the date).

**Comment by kzemek** [comment]
```
time="2017-04-14 18:05:02.911" level=info app=mongooseim pid=<0.33.0> Application mnesia started on node mongooseim@localhost
time="2017-04-14 18:05:03.217" level=notice app=mongooseim module=lager_file_backend file="/Users/konradzemek/MongooseIM/_build/default/lib/lager/src/lager_file_backend.erl" line=131 pid=<0.241.0> Changed loglevel of log/ejabberd.log to warning
time="2017-04-14 18:06:58.883" level=debug app=mongooseim jid="user_1@localhost/konradzemek" node=mongooseim@localhost pid=<0.1096.0> line=1580 function=send_text module=ejabberd_c2s application=ejabberd Send XML on stream = <<"<message from='user_1@localhost/konradzemek' to='user_1@localhost/konradzemek' xml:lang='en' type='chat' id='aac2a'>\n<composing xmlns='http://jabber.org/protocol/chatstates'/>\n</message>">>
```

**Comment by erszcz** [comment]
I have two remarks on this log format:

1. The "meta" part of each line is longer, hence log messages in general will be longer, which is not very convenient for console use. I'm not sure whether we can do anything about it.

2. It's cool that metainfo is represented as easily parseable key-value pairs. However, I don't see a clear delineation between the "meta" part and the actual log message. The message might also contain `=` signs, so we can't use them for machine processing. Maybe some extra separator could help?

File: apps/ejabberd/src/lager_elogfmt_formatter.erl [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    ?assertEqual({"level", "error"}, severity(Msg)).
+
+msg_test() ->
+    Msg = lager_msg:new("\n\t\b\r'\"\\", error, [], []),
+    ?assertEqual({"msg",
+                  ["\"", ["\\n","\\t","\\b","\\r","\\'","\\\"","\\\\"], "\""]},
+                 msg(Msg)).
+
+msg_iolist_test() ->
+    Msg = lager_msg:new(["\"", ["\n\t"],[["\b"]],"\r'\"\\"], error, [], []),
```
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 156

File: apps/ejabberd/src/lager_elogfmt_formatter.erl [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+-include_lib("eunit/include/eunit.hrl").
+
+severity_test() ->
+    Msg = lager_msg:new("msg", error, [], []),
+    ?assertEqual({"level", "error"}, severity(Msg)).
+
+msg_test() ->
+    Msg = lager_msg:new("\n\t\b\r'\"\\", error, [], []),
+    ?assertEqual({"msg",
+                  ["\"", ["\\n","\\t","\\b","\\r","\\'","\\\"","\\\\"], "\""]},
```
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 152

File: apps/ejabberd/src/lager_elogfmt_formatter.erl [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+                 msg(Msg)).
+
+msg_iolist_test() ->
+    Msg = lager_msg:new(["\"", ["\n\t"],[["\b"]],"\r'\"\\"], error, [], []),
+    ?assertEqual({"msg",
+                  ["\"", ["\\\"","\\n","\\t","\\b","\\r","\\'","\\\"","\\\\"],
```
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 158

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 1 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 156

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 2 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 152

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 3 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 158

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 2 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 152

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 3 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 158

Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 158

Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 158

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 2 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 152

File: apps/ejabberd/src/lager_elogfmt_formatter.erl [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    ?assertEqual([{["unique#", "myapp", ".", "my_unique"], 1234}],
+                 meta(Msg, "myapp")).
+
+generic_meta_atom_value_test() ->
+    Msg = lager_msg:new("msg", error, [{key, value}], []),
+    ?assertEqual([{"key", "value"}], meta(Msg, "myapp")).
+
+generic_meta_dashed_key_test() ->
+    Msg = lager_msg:new("msg", error, [{"k-e-y", "value"}], []),
+    ?assertEqual([{"k_e_y", ["\"","value","\""]}], meta(Msg, "myapp")).
```
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 204

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 2 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 152

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 3 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 158

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 2 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 152

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 3 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 158

File: apps/ejabberd/src/lager_elogfmt_formatter.erl (hunk 4 above)
Comment (gadgetci) [inline apps/ejabberd/src/lager_elogfmt_formatter.erl]: According to **Elvis**:
> Missing space right "," on line 204

**Comment by kzemek** [comment]
```
time="2017-04-19 11:52:06.627" level=info app=mongooseim msg="Application lager started on node mongooseim@localhost" pid=<0.33.0>
time="2017-04-19 11:53:04.013" level=debug app=mongooseim msg="Send XML on stream = <<\"<success xmlns=\'urn:ietf:params:xml:ns:xmpp-sasl\'/>\">>" node=mongooseim@localhost pid=<0.1109.0> line=1580 function=send_text module=ejabberd_c2s application=ejabberd
time="2017-04-19 11:53:04.013" level=info app=mongooseim msg="({socket_state,fast_tls,{tlssock,#Port<0.14885>,#Port<0.14886>},<0.1108.0>}) Accepted authentication for user_1 by ejabberd_auth_internal" node=mongooseim@localhost pid=<0.1109.0> line=3142 function=handle_sasl_success module=ejabberd_c2s application=ejabberd
```

**Comment by narg95** [comment]
Hi everyone, 

Any plans to get this merge soon?

**Comment by rslota** [comment]
> It's cool that metainfo is represented as easily parseable key-value pairs. However, I don't see a clear delineation between the "meta" part and the actual log message. The message might also contain = signs, so we can't use them for machine processing. Maybe some extra separator could help?

@erszcz , @kzemek ,
What about having `msg="Here log message"`?

**Comment by kzemek** [comment]
@rslota We'd need to escape the message; lager_elogfmt_formater module [went through this phase already](https://github.com/heroku/lager_elogfmt_formatter/commit/d1b8183207c1001c1017b9d440375a691e805db4) and I can understand why they'd back off.

**Comment by chrzaszcz** [comment]
Closing this as we are not using lager anymore.

**Pull request closed without merging**